### PR TITLE
info: improve handling of empty Info

### PR DESCRIPTION
**Pull request by thaJeztah**
- relates to https://github.com/docker/cli/issues/3593
- relates to https://github.com/docker/cli/pull/3179

Before this change, the function could print an error in some cases, for example;

```bash
docker -H tcp://127.0.0.1:2375 info --format '{{.LoggingDriver}}'

template: :1:2: executing "" at <.LoggingDriver>: reflect: indirection through nil pointer to embedded struct field Info
```

With this patch applied, the error is handled gracefully, and when failing to
connect with the daemon, the error is logged;

```bash
docker -H tcp://127.0.0.1:2375 info --format '{{.LoggingDriver}}'
Cannot connect to the Docker daemon at tcp://127.0.0.1:2375. Is the docker daemon running?

docker -H tcp://127.0.0.1:2375 info --format '{{json .}}'
Cannot connect to the Docker daemon at tcp://127.0.0.1:2375. Is the docker daemon running?
{"ID":"","Containers":0,"..."}}
```

Note that the connection error is also included in the JSON `ServerErrors` field,
so that the information does not get lost, even if STDERR would be redirected;

```bash
docker -H tcp://127.0.0.1:2375 info --format '{{json .ServerErrors}}' 2> /dev/null
["Cannot connect to the Docker daemon at tcp://127.0.0.1:2375. Is the docker daemon running?"]
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Review by rumpl — APPROVED** [review]

File: cli/command/system/info.go [hunk 1]
```diff
@@ -84,6 +85,7 @@ func runInfo(cmd *cobra.Command, dockerCli command.Cli, opts *infoOptions) error
 		if dinfo, err := dockerCli.Client().Info(ctx); err == nil {
 			info.Info = &dinfo
 		} else {
+			fmt.Fprintln(dockerCli.Err(), err)
```
Comment (rumpl) [inline cli/command/system/info.go:88]: Shouldn't we return on error?

Comment (thaJeztah) [reply cli/command/system/info.go:88]: I think this case was to allow at least the Client info to be printed (Server info would not be printed, but "failed to connect") 🤔

Comment (rumpl) [reply cli/command/system/info.go:88]: Ah right!

Comment (thaJeztah) [reply cli/command/system/info.go:88]: Gave it a quick try; here's if we don't return an error

```bash
docker -H tcp://127.0.0.1:2375 info
Cannot connect to the Docker daemon at tcp://127.0.0.1:2375. Is the docker daemon running?
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc., v0.8.2)
  compose: Docker Compose (Docker Inc., v2.4.1)
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc., 0.6.0)
  scan: Docker Scan (Docker Inc., v0.17.0)

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 0
 Plugins:
  Volume:
  Network:
  Log:
 Swarm:
  NodeID:
  Is Manager: false
  Node Address:
 CPUs: 0
 Total Memory: 0B
 Docker Root Dir:
 Debug Mode: false
 Experimental: false
 Live Restore Enabled: false

WARNING: No memory limit support
WARNING: No swap limit support
WARNING: No oom kill disable support
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
WARNING: No cpu shares support
WARNING: No cpuset support
WARNING: IPv4 forwarding is disabled
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
ERROR: Cannot connect to the Docker daemon at tcp://127.0.0.1:2375. Is the docker daemon running?
errors pretty printing info
```

And if we return the error;

```bash
docker -H tcp://127.0.0.1:2375 info
Cannot connect to the Docker daemon at tcp://127.0.0.1:2375. Is the docker daemon running?
```

Comment (thaJeztah) [reply cli/command/system/info.go:88]: From some perspective "either" can make sense, although I think it's useful to still have the Client information (even if we can't connect to the daemon), which is the same behavior as we have for `docker version`.

That said; to further improve;

- We should not print the `WARNING` output if we can't connect to the daemon (as they're inaccurate in that case). Those warnings are printed by [`printServerWarningsLegacy()`](https://github.com/docker/cli/blob/b68db383d389ae7c0302aaa3dc33f3155b6fe9a3/cli/command/system/info.go#L458-L463), which is only to be used for older daemons (and should not be called if we have no daemon to connect with).
- Similar to the `docker version` output, we should consider _not_ printing the `Server:` output if we can't connect, and instead only print the error for that part.
- ^^ perhaps an exception to the above would be if a `--format` flag is specified (?) (e.g. `--format '{{.Containers}}'`) (not sure)

**Comment by thaJeztah** [comment]
@rumpl I pushed a commit that improves the "failed to connect to server" situation; it now no longer prints the Server information (and instead only shows the error if no `--format` flag is printed)

**Review by ndeloof — APPROVED** [review]

**Pull request merged**